Wire cert contexts

ravisantoshgudimetla opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup
#### What this PR does / why we need it:
All the controllers should use context for signalling termination of communication with API server. Once kcm cancels context all the cert controllers which are started via kcm should cancel the APIServer request in flight instead of hanging around.
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

fedebongio commented:
/assign @tkashem 
/triage accepted

ravisantoshgudimetla commented:
/retest

ravisantoshgudimetla commented:
cc @damemi. Can you check if I missed anything?

damemi reviewed (commented):
@ravisantoshgudimetla one thing that I missed when doing a lot of these is that the end goal is to wire the new contexts through to all the API calls as well (so replacing every `context.TODO()`). I don't know if @soltysh still wants to do that now, or hold off for the next release. 

But just grepping through your branch I found the following that could be relevant, so you may want to check these out to make sure they trace back to the new contexts:
```
~/go/src/k8s.io/kubernetes$ grep -rni "context\.TODO" pkg/controller/certificates/*
pkg/controller/certificates/approver/sarapprove.go:135:	sar, err := a.client.AuthorizationV1().SubjectAccessReviews().Create(context.TODO(), sar, metav1.CreateOptions{})
pkg/controller/certificates/certificate_controller_test.go:51:		_, err := client.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.TODO(), csr.Name, csr, metav1.UpdateOptions{})
pkg/controller/certificates/cleaner/cleaner.go:108:		if err := ccc.csrClient.Delete(context.TODO(), csr.Name, metav1.DeleteOptions{}); err != nil {
pkg/controller/certificates/rootcacertpublisher/publisher.go:192:		_, err = c.client.CoreV1().ConfigMaps(ns).Create(context.TODO(), &v1.ConfigMap{
pkg/controller/certificates/rootcacertpublisher/publisher.go:227:	_, err = c.client.CoreV1().ConfigMaps(ns).Update(context.TODO(), cm, metav1.UpdateOptions{})
pkg/controller/certificates/signer/signer.go:170:		_, err = s.client.CertificatesV1().CertificateSigningRequests().UpdateStatus(context.TODO(), csr, metav1.UpdateOptions{})
pkg/controller/certificates/signer/signer.go:184:	_, err = s.client.CertificatesV1().CertificateSigningRequests().UpdateStatus(context.TODO(), csr, metav1.UpdateOptions{})

~/go/src/k8s.io/kubernetes$ grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/*
staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller.go:187:	configMap, err := c.client.CoreV1().ConfigMaps(c.configmapNamespace).Get(context.TODO(), c.configmapName, metav1.GetOptions{})

~/go/src/k8s.io/kubernetes$ grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/server/options/*
staging/src/k8s.io/apiserver/pkg/server/options/serving.go:360:	ln, err := config.Listen(context.TODO(), network, addr)

~/go/src/k8s.io/kubernetes$ grep -rni "context\.TODO" staging/src/k8s.io/kube-aggregator/pkg/apiserver/*
staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:307:					context.TODO(), s.GenericAPIServer.APIServerID, metav1.GetOptions{})
```

(note there could be more, but I just searched through the folders you've already edited)

ravisantoshgudimetla commented:
@damemi - Thanks for the review. I was initially thinking the same. However as you might have already noticed that - if the context is cancelled, Run anyway is going to stop, so rest of the methods even if they're using the context.TODO() may not have effect. To me passing `ctx` to New methods did not make much sense. I wish the controller's handlers are decoupled from `CertificateController` [struct](https://github.com/kubernetes/kubernetes/blob/704dc9a99e6a499a8b36d4637b99b63960e36837/pkg/controller/certificates/certificate_controller.go#L53-L52), so that we can call them via `Run` so that we have a single entry point and no initialization. For example, we can make CertificateController an interface and Run method there.

I did not want to make all the changes in a single PR and make it hard to review. I can comeback to this later if you and @soltysh are fine with the approach I suggested.

soltysh commented:
@ravisantoshgudimetla  `Run` won't be cancelled if it hangs on a client call, that's why we need to pass the context all the way down to all client invocation. I'm looking at the PR and I'm wondering if we could split it into smaller bits, like just the apiserver bits and necessary places using `DynaminFileCAContent` and split controllers along with the changes requested by @damemi in a followup?

aojea commented on cmd/kubelet/app/auth.go:
```diff
@@ -95,8 +96,18 @@ func BuildAuthn(client authenticationclient.AuthenticationV1Interface, authn kub
+		// TODO: See if we can pass ctx to the current method
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			select {
+			case <-stopCh:
+				cancel() // stopCh closed, so cancel our context
+			case <-ctx.Done():
+			}
```
I'm not able to understand well this pattern, does this means that the context can be cancelled from the inside?

ravisantoshgudimetla replied:
Yeah, the goal is to have a routine which looks at the stopChannel and cancels the context when stopChannel is closed.

tkashem replied:
> yeah, but AFAIK the context is meant to be cancelled from the parent not from the childs,

yes, this is the recommended use, but here the lifecycle aspect of the context is scoped within the go routine, it's because only when the `stopCh` is closed should we cancel the context. If we return the cancel func to the caller this is no longer guaranteed.

I would recommend:
```
go func() {
                        defer cancel()
			select {
			case <-stopCh:
			case <-ctx.Done():
			}
           }
```

tkashem replied:
i take it back, we don't need a defer in this case.

aojea replied:
you are only blocking on the stopCh to cancel the context
```
		ctx, cancel := context.WithCancel(context.Background())
		go func() {
			<-stopCh
			cancel() // stopCh closed, so cancel our context
			}
```
the only way this goes out is if someone cancels the context, but that will mind that it has to be done from within the children

ravisantoshgudimetla commented:
@damemi Not sure if you looked at the right branch but most of the changes should be already in place. I was suggesting how it be better in the long run.

```
➜  kubernetes git:(wire-contexts-certificates) grep -rni "context\.TODO" pkg/controller/certificates/*
pkg/controller/certificates/approver/sarapprove_test.go:134:			ctx := context.TODO()
pkg/controller/certificates/cleaner/cleaner_test.go:230:			ctx := context.TODO()
pkg/controller/certificates/signer/signer_test.go:296:			ctx := context.TODO()
➜  kubernetes git:(wire-contexts-certificates) grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/*
staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go:224:			ctx := context.TODO()
```

ravisantoshgudimetla commented:
Made some changes after picking Mike's commits:

```
➜  kubernetes git:(wire-cert-contexts) ✗ grep -rni "context\.TODO" pkg/controller/certificates/*
pkg/controller/certificates/approver/sarapprove_test.go:134:			ctx := context.TODO()
pkg/controller/certificates/certificate_controller_test.go:50:		_, err := client.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.TODO(), csr.Name, csr, metav1.UpdateOptions{})
pkg/controller/certificates/certificate_controller_test.go:72:	ctx := context.TODO()
pkg/controller/certificates/cleaner/cleaner_test.go:229:			ctx := context.TODO()
pkg/controller/certificates/rootcacertpublisher/publisher_test.go:158:			ctx := context.TODO()
pkg/controller/certificates/rootcacertpublisher/publisher_test.go:267:			ctx := context.TODO()
pkg/controller/certificates/signer/signer_test.go:298:			ctx := context.TODO()

========================================================================


➜  kubernetes git:(wire-cert-contexts) ✗ grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/*
staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go:225:			ctx := context.TODO()
=========================================================================


➜  kubernetes git:(wire-cert-contexts) ✗ grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/server/options/*
staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:393:	ctx := context.TODO()
staging/src/k8s.io/apiserver/pkg/server/options/serving.go:360:	ln, err := config.Listen(context.TODO(), network, addr)
➜  kubernetes git:(wire-cert-contexts) ✗ grep -rni "context\.TODO" staging/src/k8s.io/kube-aggregator/pkg/apiserver/*
staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:249:		ctx := context.TODO()
staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:320:					context.TODO(), s.GenericAPIServer.APIServerID, metav1.GetOptions{})

We need to pass a dummy context but the controller is not making any api calls.
```

I don't mind splitting the PR's commits into 2 different commits. Let me know if  it makes reviews easier.

ravisantoshgudimetla committed:
API Server Changes

This commit includes all the changes needed for APIServer. Instead of modifying the existing signatures for the methods which either generate or return stopChannel, we generate a context from the channel and use the generated context to be passed to the controllers which are started in APIServer. This ensures we don't have to touch APIServer dependencies.

ravisantoshgudimetla committed:
Wire context for cert controllers

All the controllers should use context for signalling termination of communication with API server. Once kcm cancels context all the cert controllers which are started via kcm should cancel the APIServer request in flight instead of hanging around.

tkashem commented on staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go:
```diff
@@ -221,7 +222,8 @@ func TestRequestHeaderAuthRequestControllerSyncOnce(t *testing.T) {
 			// act
-			err := target.RunOnce()
+			ctx := context.TODO()
+			err := target.RunOnce(ctx)
```
nit: `target.RunOnce(context.TODO())`

tkashem commented on staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:
```diff
@@ -387,7 +388,10 @@ func (s *DelegatingAuthenticationOptions) createRequestHeaderConfig(client kuber
+	//  We are passing the context to ProxyCerts.RunOnce as it needs to implement RunOnce(ctx) however the
+	//  context is not used at all. So passing a empty context shouldn't be a problem
+	ctx := context.TODO()
+	if err := dynamicRequestHeaderProvider.RunOnce(ctx); err != nil {
```
nit:  `dynamicRequestHeaderProvider.RunOnce(context.TODO())`

tkashem commented on pkg/controlplane/instance.go:
```diff
@@ -452,29 +452,40 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
+		// generate a context  from stopCh. This is to avoid modifying files which are relying on apiserver
+		// TODO: See if we can pass ctx to the current method
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
```
nit: is there any common place where we can define this function and use it here?

tkashem reviewed (commented):
LGTM (just have a few minor nits)

sttts commented:
/lgtm
/approve

soltysh reviewed (approved):
/lgtm
/approve
/release-note none
/kind feature

soltysh commented:
/release-note-none

rphillips commented:
PR description needs more context... Looks like a cleanup PR?

rphillips commented:
lgtm after that update.

rphillips commented:
/retest

ravisantoshgudimetla commented:
/retest
